storage: don't mutate client's protos

tbg opened the pull request:
Prior to fix, would fail with

```
--- FAIL: TestReplicaTimestampCacheBumpNotLost (0.00s)
       	replica_test.go:6282: original transaction proto was mutated: [TxnMeta.Timestamp.Logical: 20 != 22]
```

Also added an assertion against the following potential anomaly (tracked as
issue #10401):

- client starts txn at ts=1
- client writes and applyTimestampCache bumps the write to ts=2
- client receives no response with its request, keeps thinking ts=1 is the
  current timestamp
- client commits at ts=1
- one of its writes is at ts=2, and there is no correct way of resolving the
  situation; either the write is lost or it commits, invalidating a prior read.

Unfortunately, that assertion fires and it was downgraded to a workaround plus
a Warning. Obviously, it should be addressed.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10402)
<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1364 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaT3vramVFCMH908z:-KVaaT3vramVFCMH909-:b-bq5grx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica.go#L1364)):*
> ```Go
> 	defer func() {
> 		// Let the caller know whether we did anything.
> 		bumped = (ba.Txn == nil && origTS != ba.Timestamp) ||
> ```

```
if ba.Txn != nil {
  bumped = origTS != ba.Txn.Timestamp
} else {
  bumped = origTS != ba.Timestamp
}
```
This reads more easily and nicely mirrors the above.

---

*[pkg/storage/replica.go, line 1379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaq6dX_nPsylc1DiJ:-KVaaq6dX_nPsylc1DiK:b-otxczn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica.go#L1379)):*
> ```Go
> 				wTS, _, wOK := r.mu.tsCache.GetMaxWrite(key, nil)
> 				if wOK {
> 					return false, roachpb.NewError(roachpb.NewTransactionReplayError())
> ```

throughout: `return bumped`? `false` here is meaningless since `bumped` is entirely evaluated in the defer

---

*[pkg/storage/replica_test.go, line 6244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaa2mtt9i8CB0CBbWG:-KVaa2mtt9i8CB0CBbWH:b-lptdpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6244)):*
> ```Go
> 	defer tc.Stop()
> 
> 	ctx, sp := tc.store.AnnotateCtxWithSpan(context.TODO(), "test")
> ```

the span is required?

---

*[pkg/storage/replica_test.go, line 6263 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaa5kT2P535K6EAhwz:-KVaa5kT2P535K6EAhx-:bsyerlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6263)):*
> ```Go
> 		}
> 		if !txn.Timestamp.Less(resp.Timestamp) {
> 			t.Fatal("expected txn ts %s < scan TS %s", txn.Timestamp, resp.Timestamp)
> ```

Fatalf

---

*[pkg/storage/replica_test.go, line 6281 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaDVa712hVSB8G6pI:-KVaaDVa712hVSB8G6pJ:boqs7dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6281)):*
> ```Go
> 		t.Fatalf(
> 			"original transaction proto was mutated: %s",
> 			pretty.Diff(&origTxn, txn),
> ```

nit: this is backwards from the DeepEqual call, tickling my OCD

---

*[pkg/storage/replica_test.go, line 6285 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaI06cxJxk3uUwLq9:-KVaaI06cxJxk3uUwLqA:b-xn608x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6285)):*
> ```Go
> 	}
> 	if resp.Txn == nil {
> 		t.Fatal("no response in transaction")
> ```

no transaction in response?

---

*[pkg/storage/replica_test.go, line 6287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaGj-XzmfwlpkbTfG:-KVaaGj-XzmfwlpkbTfH:bsyerlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6287)):*
> ```Go
> 		t.Fatal("no response in transaction")
> 	} else if resp.Txn.Timestamp.Less(minNewTS) {
> 		t.Fatal(
> ```

Fatalf

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: don't mutate client's protos

Prior to fix, would fail with

```
--- FAIL: TestReplicaTimestampCacheBumpNotLost (0.00s)
       	replica_test.go:6282: original transaction proto was mutated: [TxnMeta.Timestamp.Logical: 20 != 22]
```

Also added an assertion against the following potential anomaly (tracked as
issue #10401):

- client starts txn at ts=1
- client writes and applyTimestampCache bumps the write to ts=2
- client receives no response with its request, keeps thinking ts=1 is the
  current timestamp
- client commits at ts=1
- one of its writes is at ts=2, and there is no correct way of resolving the
  situation; either the write is lost or it commits, invalidating a prior read.

Unfortunately, that assertion fires and it was downgraded to a workaround plus
a Warning. Obviously, it should be addressed

tschottdorf commented:
TFTR, comments addressed.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1364 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaT3vramVFCMH908z:-KVad1ZDJv_SX8cTfbMF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica.go#L1364)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ```
> if ba.Txn != nil {
>   bumped = origTS != ba.Txn.Timestamp
> } else {
>   bumped = origTS != ba.Timestamp
> }
> ```
> This reads more easily and nicely mirrors the above.
</details>
Done.

---

*[pkg/storage/replica.go, line 1379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaq6dX_nPsylc1DiJ:-KVad6j6_0u8_BrjJxM8:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica.go#L1379)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: `return bumped`? `false` here is meaningless since `bumped` is entirely evaluated in the defer
</details>
Good point, done.

---

*[pkg/storage/replica_test.go, line 6244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaa2mtt9i8CB0CBbWG:-KVadL3zn-hmmRlXhi0L:bl23n7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6244)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the span is required?
</details>
Nope, autocompleted incorrectly. Removed.

---

*[pkg/storage/replica_test.go, line 6263 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaa5kT2P535K6EAhwz:-KVadQhAQYwzqRFN5Bxl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6263)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Fatalf
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6281 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaDVa712hVSB8G6pI:-KVadWoEkFtK27z-PIwc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6281)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: this is backwards from the DeepEqual call, tickling my OCD
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6285 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaI06cxJxk3uUwLq9:-KVadTfwtRbuJefoM4Qp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6285)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> no transaction in response?
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-KVaaGj-XzmfwlpkbTfG:-KVadULZMneE69tpde8Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a690014a736b273b0284f413cc5186f5185402b/pkg/storage/replica_test.go#L6287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Fatalf
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-:-KVaeWex-EhPqeBBCQKO:b-dafajm)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10402#-:-KVafa9vy5K5fIe2JmI0:bnfp4nl)*
<!-- Sent from Reviewable.io -->
